jerry-core/ecma/builtin-objects/ecma-builtin-global.c

```diff
@@ -510,76 +545,39 @@ ecma_builtin_global_object_unescape (lit_utf8_byte_t *input_start_p, /**< routin
                                      lit_utf8_size_t input_size) /**< routine's first argument's
                                                                   *   string buffer's size */
 {
+  if (input_size == 0)
+  {
+    return ecma_make_magic_string_value (LIT_MAGIC_STRING__EMPTY);
+  }
+
   const lit_utf8_byte_t *input_curr_p = input_start_p;
   const lit_utf8_byte_t *input_end_p = input_start_p + input_size;
-  /* 4. */
-  /* The length of input string is always greater than output string
-   * so we re-use the input string buffer.
-   * The %xx is three byte long, and the maximum encoded value is 0xff,
-   * which maximum encoded length is two byte. Similar to this, the maximum
-   * encoded length of %uxxxx is four byte. */
-  lit_utf8_byte_t *output_char_p = input_start_p;
-
-  /* The state of parsing that tells us where we are in an escape pattern.
-   * 0    we are outside of pattern,
-   * 1    found '%', start of pattern,
-   * 2    found first hex digit of '%xy' pattern
-   * 3    found valid '%xy' pattern
-   * 4    found 'u', start of '%uwxyz' pattern
-   * 5-7  found hex digits of '%uwxyz' pattern
-   * 8    found valid '%uwxyz' pattern
-   */
-  uint8_t status = 0;
-  ecma_char_t hex_digits = 0;
-  /* 5. */
+  ecma_stringbuilder_t builder = ecma_stringbuilder_create ();
+
   while (input_curr_p < input_end_p)
   {
-    /* 6. */
     ecma_char_t chr = lit_cesu8_read_next (&input_curr_p);
+    lit_utf8_size_t byte_offset = (lit_utf8_size_t) (input_end_p - input_curr_p);
 
-    /* 7-8. */
-    if (status == 0 && chr == LIT_CHAR_PERCENT)
-    {
-      /* Found '%' char, start of escape sequence. */
-      status = 1;
-    }
-    /* 9-10. */
-    else if (status == 1 && chr == LIT_CHAR_LOWERCASE_U)
-    {
-      /* Found 'u' char after '%'. */
-      status = 4;
-    }
-    else if (status > 0 && lit_char_is_hex_digit (chr))
-    {
-      /* Found hexadecimal digit in escape sequence. */
-      hex_digits = (ecma_char_t) (hex_digits * 16 + (ecma_char_t) lit_char_hex_to_int (chr));
-      status++;
-    }
-    else
-    {
-      /* Previously found hexadecimal digit in escape sequence but it's not valid '%xy' pattern
-       * so essentially it was only a simple character. */
-      status = 0;
-    }
-
-    /* 11-17. Found valid '%uwxyz' or '%xy' escape. */
-    if (status == 8 || status == 3)
+    // potential pattern
+    if (chr == LIT_CHAR_PERCENT)
     {
-      output_char_p -= (status == 3) ? 2 : 5;
-      status = 0;
-      chr = hex_digits;
-      hex_digits = 0;
+      // potential unicode sequence
+      if (byte_offset >= 5 && input_curr_p[0] == LIT_CHAR_LOWERCASE_U)
+      {
+        ecma_builtin_global_object_unescape_helper (&input_curr_p, &chr, 5, true);
```

Comment:
Why don't we increment `input_curr_p` before calling the unescape helper, the there's no need for the `unicode_sequence` argument.